client/state.cpp

```diff
@@ -229,38 +226,34 @@ std::vector<std::string> State::update(
   return upd;
 }
 
-bool State::update_frame(const torchcraft::fbs::FrameData* fd) {
-  if (fd->data() == nullptr)
-    return false;
-  if (fd->is_diff()) {
-    this->frame_string = "";
-    std::istringstream ss(std::string(
-        reinterpret_cast<const char*>(fd->data()->data()), fd->data()->size()));
-    replayer::FrameDiff diff;
-    ss >> diff;
-    replayer::frame_undiff(this->frame, this->frame, &diff);
-  } else {
-    this->frame_string = std::string(
-        reinterpret_cast<const char*>(fd->data()->data()), fd->data()->size());
-    std::istringstream ss(this->frame_string);
-    ss >> *this->frame;
+bool State::update_frame(const void* frameOrFrameDiff) {
+
+  auto asFrame = static_cast<const fbs::Frame*>(frameOrFrameDiff);
```

Comment:
This won't work; `Client::receive()` already contains some branch based on the actual type of the message.

Reply:
I think we're okay. Client::receive() distinguishes between message types but not Frame/FrameDiff. It passes the FrameState (Formerly "Frame") which contains the Frame/FrameDiff union through to State::update() which then invokes update_frame() without any branch.

Reply:
But this looks like you're relying on `static_cast` to result in `nullptr` for non-matching types -- that's not how it works. Normally, `dynamic_cast` can be used for this but only if `Frame` or `FrameDiff` were part of the same hierarchy and the void pointer you're providing to the function is part of it as well (IIRC there are some limitations to what kind of casts work with `dynamic_cast` even within a hierarchy).

Reply:
What I meant with the reference to `Client::receive()` is that it contains code to distinguish between different types in a union, which is exactly what you need here.